### DISCUSSION TITLE
Reading values from properties file on karavan.

### DISCUSSION THREAD

**Discussion by MN-Dheeraj** [apache/camel-karavan#1242] [Q&A] [answered]
### Ask a question

I have code on karavan that needs values from environment file(properties). Is it possible? If yes, how do I that?

### Variant

Web Application

### Container Management (if applicable)

Docker

### Operating System (if applicable)

Windows

### Version

4.4.0

**Comment by mgubaidullin** [comment]
See property handling part https://camel.apache.org/blog/2024/03/camel-karavan-4.4.0/

**Reply by MN-Dheeraj** [reply]
Hi @mgubaidullin , Thanks for the reply.
In my case, I need to read it in my java code in processors. I already tried injecting variable using {{key}}. But, It didn't work out.
Since, I can't use spring.Environment, I tried this way.

May I know if there is a way to use that key-value pairs of properties in my code?

**Reply by davsclaus** [reply]
https://camel.apache.org/manual/using-propertyplaceholder.html#_resolving_property_placeholders_from_java_code

And you can use `@PropertyInject`
https://camel.apache.org/manual/bean-integration.html

**Reply by lalyko** [reply]
Hi Karavan-Team,
 in my project I'm facing the same issue. The suggested solutions by davsclaus didn't work.
By example, for a https component, I can select the properties from the application.properties file and the property ist written to the text field (Http Uri) of the component, but the run console shows an exception.

![image](https://github.com/apache/camel-karavan/assets/85667665/8d60c8ba-de3d-442f-a124-7edf1bf548cb)
![image](https://github.com/apache/camel-karavan/assets/85667665/bdfa06e6-a0bc-4674-9b69-04caa5d2ecf8)
![image](https://github.com/apache/camel-karavan/assets/85667665/2afd41b2-8cdc-4477-8dad-b89ba457dab2)

**Reply by mgubaidullin** [reply]
Http component is not dynamic. You can not use application.properties in component properties.

Set headers used by Http component instead

**Reply by davsclaus** [reply]
Oh its not dynamic values per message but property placeholders where you put the values into application.properties. That should always be possible anywhere in the DSL.

It may be that the karavan dev mode somehow does not have application.properties updated when starting. If you try with pure camel jbang then it ought to work

**Reply by MN-Dheeraj** [reply]
In my case, these are the observations, @davsclaus 

1st option:
Using @PropertyInject("key-name") doesn't work for classes which are not processors. The values from application.properties could be injected in processors on routes. Can you please confirm if this is expected?

2nd option:

import java.util.Optional;
import org.apache.camel.CamelContext;
import org.apache.camel.impl.DefaultCamelContext;
import org.apache.camel.component.properties.PropertiesComponent;
CamelContext camelContext = new DefaultCamelContext();
PropertiesComponent propertiesComponent = new PropertiesComponent();
propertiesComponent.setLocation("org/camel/karavan/demo/il/lc/application.properties");
camelContext.setPropertiesComponent(propertiesComponent);

Optional<String> prop = camelContext.getPropertiesComponent().resolveProperty("logging.enable");
if (prop.isPresent()) {
	String value = prop.get();
	System.out.println("context logEnable = " + value);
}

The above one doesn't work at all. The value of 'logging.enable' from application.properties couldn't be injected. I was getting a null value.
Can you please let me now if the above logic is not correct?

Best Regards,
MN. Dheeraj

**Reply by lalyko** [reply]
Even using @PropertyInject("custom.prop.x") doesn't work for processor classes!
@davsclaus - how can I fix reading custom properties of the appication.properties file in dev mode?

Best regards,
R. Lalyko

**Reply by MN-Dheeraj** [reply]
@lalyko , I hope you added the correct package of PropertyInject in your code. I also added a custom property in my application.properties and injected it using @PropertyInject. It worked for me actually.

I also added this property in my application.properties. However, I am not sure if the below property is mandatory.
camel.component.properties.auto-discover-properties-sources=true

**Comment by lalyko** [comment]
@MN-Dheeraj , in my processor I've imported org.apache.camel.PropertyInject and it doesn' work ... see my screenshots

![image](https://github.com/apache/camel-karavan/assets/85667665/e87c5274-442c-4261-b1cb-0aa4b6de0165)
![image](https://github.com/apache/camel-karavan/assets/85667665/6c81a52c-5fdc-498f-96e3-1bac220c0b04)
![image](https://github.com/apache/camel-karavan/assets/85667665/d4798e0c-9768-4c42-b7c8-d6bac8cfc85c)

**Comment by lalyko** [comment]
@MN-Dheeraj , even resolving custom properties form camelContext does not work ...

![image](https://github.com/apache/camel-karavan/assets/85667665/b98bd9de-0310-41cd-a369-72f9f4bdb299)
![image](https://github.com/apache/camel-karavan/assets/85667665/fba8dc96-f482-421d-a7c5-d1c3a5f93f09)

**Reply by MN-Dheeraj** [reply]
Oh, resolving from properties component didn't work for me either.
However, I am not sure why injecting in processor using @PropertyInject didn't work. 
It seems we are missing something and both of us will need some help.

**Comment by lalyko** [comment]
@davsclaus - in my environment the cuomization of docker-compose.yaml doesn't work also. It seems karavan doesn't use the application.properties and docker-ompose.yaml for the jbang generation.

**Comment by lalyko** [comment]
@MN-Dheeraj  - as @davsclaus said in his previous post, application.properties should work for container builds and deployment, but not in the devmode (direct run in karavan). 
To use @PropertyInject you have to set a defaultValue to your properties! 
For testing, I've defined a simple TestProcessor reading the property form a properties file (1), loading the property by PropertyInject(2) and loading the property by PropertiesComponent (3).
![image](https://github.com/apache/camel-karavan/assets/85667665/291fbf44-ab58-4f33-9398-51fc49658431)

You use your own property file by adding the file by jbang files declaration: 
//FILES application-dev.properties=/karavan/code/application-dev.properties
But you have to manage the properties on your own ...
[TestProcessor.java.txt](https://github.com/apache/camel-karavan/files/15151418/TestProcessor.java.txt)

I tried to use the jbang files declaration to provide the application.properties for the devmode.
(... to run my route in karavan directly)
BUT, it doesn't work! ... even not for reading as resource (1)
![image](https://github.com/apache/camel-karavan/assets/85667665/eac1138d-18ad-4864-83eb-511f15648024)

@davsclaus - may the karavan team can sole this issue?  

Best regards,
R. Lalyko

**Reply by davsclaus** [reply, marked as answer]
Should be fixed by
https://issues.apache.org/jira/browse/CAMEL-20672

**Reply by MN-Dheeraj** [reply]
@lalyko , I get you. I actually don't have any code in src folder. All the files(java, properties etc) are in the project's root folder.
Here is my code.
Actually, I took this context not from the exchange. That was my problem. I modified this as per your code and it worked for me. Thanks.

Optional<String> prop = exchange.getContext().getPropertiesComponent().resolveProperty("logging.enable");
		String loggingEnable = "";
		if (prop.isPresent()) {
			loggingEnable = prop.get();
			System.out.println("context logEnable = " + loggingEnable);
		}

**appplication.properties:**
#Util properties#
logging.enable=true

**Startup-Logs:**
2024-04-30 10:54:02.717  INFO 18716 --- [           main] g.apache.camel.main.BaseMainSupport : Properties location: file:application.properties
2024-04-30 10:54:02.771  INFO 18716 --- [           main] apache.camel.main.ProfileConfigurer : The application is starting with profile: dev
2024-04-30 10:54:12.467  INFO 18716 --- [           main] g.apache.camel.main.BaseMainSupport : Auto-configuration summary
...other-logs..
2024-04-30 10:54:12.906  INFO 18716 --- [           main] mel.cli.connector.LocalCliConnector : Management from Camel JBang enabled
2024-04-30 10:54:13.034  INFO 18716 --- [           main] el.impl.engine.AbstractCamelContext : Apache Camel 4.5.0 (Apiconstants) is starting
2024-04-30 10:54:13.560  INFO 18716 --- [           main] .core.spi.resolver.ResolverProvider : Using the default address resolver as the dns resolver could not be loaded
2024-04-30 10:54:13.706  INFO 18716 --- [ntloop-thread-0] .http.vertx.VertxPlatformHttpServer : Vert.x HttpServer started on 0.0.0.0:8080
2024-04-30 10:54:13.791  INFO 18716 --- [           main] g.apache.camel.main.BaseMainSupport : Property-placeholders summary
2024-04-30 10:54:13.792  INFO 18716 --- [           main] g.apache.camel.main.BaseMainSupport :     [application.properties]       logging.enable=true
2024-04-30 10:54:13.907  INFO 18716 --- [           main] el.impl.engine.AbstractCamelContext : Routes startup (total:1)2024-04-30 10:54:13.907  INFO 18716 --- [           main] el.impl.engine.AbstractCamelContext :     Started route-a27c (rest://post:/LoanAccountDetails)
